File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/happy_number.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// O(logn) / O(1)
bool is_happy(int n) {
  auto sum_sqr_digits = [](int m) {
    int s = 0;
    while (m) {
      s += (m % 10) * (m % 10);
      m /= 10;
    }

    return s;
  };

  int slow = n, fast = n;
  do {
    slow = sum_sqr_digits(slow);
    fast = sum_sqr_digits(sum_sqr_digits(fast));
  } while (slow != fast);

  return slow == 1;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;
    cout << (is_happy(n) ? "true" : "false") << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/HappyNumber.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class HappyNumber {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        System.out.println(isHappy(n));
      }
    }
  }

  // O(logn) / O(1)
  public static boolean isHappy(int n) {
    int slow = n, fast = n;

    Function<Integer, Integer> sumSqrDigits = (m) -> {
      int s = 0;
      while (m > 0) {
        s += (m % 10) * (m % 10);
        m /= 10;
      }
      return s;
    };

    do {
      slow = sumSqrDigits.apply(slow);
      fast = sumSqrDigits.apply(sumSqrDigits.apply(fast));
    } while (slow != fast);

    return slow == 1;
  }
}